backport of changes in #17057 that don't require new worlds

vtjnash opened the pull request:
Fixes some method definition orderings, and other `@generated` functions, that will be broken by PR #17057

vtjnash committed:
make inference more inferrable (aka faster)

vtjnash committed:
throw actual error message for broken generated functions

ref #18605

vtjnash committed:
fix many generated functions definition ordering

vtjnash committed:
avoid runtime @eval in @parallel test

vtjnash committed:
make singleton objects more inferable and remove invalid @generated functions from irrationals

vtjnash committed:
fixup rebase error in "split SourceInfo out of LambdaInfo"

ref fix #18404

vtjnash committed:
correct and update docs for defining generated functions

tkelman commented on base/inference.jl:
```diff
@@ -3923,10 +3931,29 @@ end
+        end
+    end
+    for f in fs
+        for m in _methods_by_ftype(Tuple{typeof(f), Vararg{Any}}, 10)
```
should this 10 be given a name?

vtjnash replied:
it's just a random round-ish number

tkelman commented on base/irrationals.jl:
```diff
@@ -57,27 +57,44 @@ end
-    by < 0 && T <: Unsigned && return false
-    ry = rationalize(T,by,tol=0)
-    ry < by ? :(x <= $ry) : :(x < $ry)
+@pure function rationalize{T<:Integer}(::Type{T}, x::Irrational)
```
this causes a method overwrite warning, why is it needed?

vtjnash replied:
for the same reason the other method overwrite warning happens in searchsorted

tkelman replied:
any horizon for not needing that?

JeffBezanson replied:
The one in this case might not be needed --- calling a kwarg function with no keyword arguments is fast. The problem in searchsorted was that computing the keyword arg default values required dynamic dispatch.

vtjnash replied:
True, the problem in this case is actually just that lowering won't apply `@pure` to the kwargs functions, and that's quite fixable.

tkelman replied:
add a TODO comment saying to fix pure for kwargs so the no-kwargs method can be deleted then?

vtjnash replied:
original issue should be corrected by 3f2213f4e74ed3070045da139f659694b7c071a8 now

tkelman commented on doc/manual/metaprogramming.rst:
```diff
@@ -958,31 +967,35 @@ used?
+After that, for this example, the expression returned from the generated
+function on the first invocation was re-used as the method body.
+The actual caching behavior is an implementation-defined performance optimization,
+so it is invalid to depend to closely on this behavior, however.
```
"too closely"

don't end a sentence with ", however." - just delete that

tkelman commented on doc/manual/metaprogramming.rst:
```diff
@@ -958,31 +967,35 @@ used?
+function ``foo(x) = x * x`` could not do (except printing the type on the
+first invocation, and incurring higher overhead).
+However, the power of a generated function lies in its ability to compute
+different quoted expression depending on the types passed to it:
```
different quoted expressions

vtjnash commented on doc/manual/metaprogramming.rst:
```diff
@@ -958,31 +967,35 @@ used?
 
 The example generated function ``foo`` above did not do anything a normal
-function ``foo(x)=x*x`` could not do, except printing the type on the
-first invocation (and incurring a higher compile-time cost). However, the
```
no, it's adds overhead in many places

vtjnash committed:
lower prologue attributes to the kwarg function definition sans keywords

mauro3 commented on doc/manual/metaprogramming.rst:
```diff
@@ -893,37 +893,46 @@ the arguments are known, but the function is not yet compiled.
     julia> @generated function foo(x)
-               println(x)
-               return :(x*x)
+               Core.println(x)
```
But this is not pure.  This is mentioned way later, but coming after the new point "4." this is just odd.  Maybe state that doing `println` is ok?

vtjnash replied:
`println` isn't OK, it's just a convenient function for showing how they work (and along the way demonstrating why it isn't OK)

mauro3 commented on doc/manual/metaprogramming.rst:
```diff
@@ -893,37 +893,46 @@ the arguments are known, but the function is not yet compiled.
 3. Instead of calculating something or performing some action, you return
    a *quoted expression* which, when evaluated, does what you want.
 
+4. The function must not have any side-effects or examine any non-constant
```
To go with the later "they"s, start with "4. Generated functions must not..."

mauro3 commented on doc/manual/metaprogramming.rst:
```diff
@@ -893,37 +893,46 @@ the arguments are known, but the function is not yet compiled.
 2. In the body of the generated function you only have access to the
-   *types* of the arguments, not their values.
+   *types* of the arguments -- not their values -- and any function that
+   was defined *before* the definition of generated function.
```
"was defined *before* the definition of the generated function."

Sacha0 commented on doc/manual/metaprogramming.rst:
```diff
@@ -1028,12 +1037,40 @@ copy them*, for the following reasons:
 * the ``foo`` function has side-effects, and it is undefined exactly when,
```
As `@mauro3` suggested in https://github.com/JuliaLang/julia/pull/18679/files#r80905058, perhaps also mention the issues immediately alongside the corresponding examples?

Sacha0 commented on doc/manual/metaprogramming.rst:
```diff
@@ -1028,12 +1037,40 @@ copy them*, for the following reasons:
+     This condition is relaxed for incrementally-loaded precompiled modules to
+     allow calling any function in the module.
+
+Alright, now that we have a better understanding for how generated functions
```
for -> of ("now that we have a better understanding of how generated functions work").

tkelman commented on doc/manual/metaprogramming.rst:
```diff
@@ -1034,7 +1034,7 @@ These examples are hopefully helpful to illustrate how generated functions
 copy them*, for the following reasons:
 
-* the ``foo`` function has side-effects, and it is undefined exactly when,
+* the ``foo`` function has side-effects (the call to `Core.println`), and it is undefined exactly when,
```
rst, so double backticks

vtjnash committed:
fix a slottypes bug exposed by making singleton objects more inferable

the widenconst for slottypes happens before optimization,
so need to be careful not to introduce `Const` in that array during optimization

vtjnash commented:
rebased to squash (previous head was 8f4d569) and added 94be098bbedb86eb575b17067fa5ded173eff3a8 to fix a pre-existing bug, for backporting (not all of these commits need to be backported, but can review that later)

tkelman commented:
@nanosoldier `runbenchmarks(ALL, vs=:master)`

jrevels commented:
@nanosoldier `runbenchmarks(ALL, vs=":master")`

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/18679#issuecomment-251160136) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/ac007972fd8a729bb107c27c47912e083e6de484/1560b17_vs_6c27e24/report.md). cc @jrevels

vtjnash commented:
nanosoldier says, "merge"?

maleadt commented on src/gf.c:
```diff
@@ -829,7 +829,7 @@ static jl_method_instance_t *jl_mt_assoc_by_type(jl_methtable_t *mt, jl_datatype
     jl_method_instance_t *nf;
     if (!cache) {
-        nf = jl_get_specialized(m, sig, env); // TODO: should be jl_specializations_get_linfo
+        nf = jl_specializations_get_linfo(m, sig, env);
```
Just FYI, this change doesn't fix the ASAN issue I mentioned.

KristofferC commented:
I'm excited for the new worlds to come.
